### DISCUSSION TITLE
插入空值会导致服务挂，起不起来，只能全量删除数据，重新部署吗

### DISCUSSION THREAD

**Discussion by yangyangwen** [milvus-io/milvus#20952] [Q&A and General discussion]
使用java-sdk写入如果是空字符串的数据，服务会直接挂掉，起也起不起来，报错显示节点刷磁盘报错。有解决办法吗？[datanode/flow_graph_insert_buffer_node.go:402] ["insertBufferNode flushBufferData failed, err = All attempts results:\nattempt #1:can't add empty string into payload\nattempt #2:can't add empty string into payload\nattempt #3:can't add empty string into payload\nattempt #4:can't add empty string into payload\nattempt #5:can't add empty string into payload\n"] [stack="github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Operate\n\t/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_insert_buffer_node.go:402\ngithub.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work\n\t/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:102"]

**Comment by yhmo** [comment]
是2.1.2或更早的版本么？是有这样的问题，在2.1.3里修复的。
https://github.com/milvus-io/milvus/issues/18872
https://github.com/milvus-io/embd-milvus/issues/9

**Reply by yangyangwen** [reply]
是的，使用的是2.1版本的，好的感谢。另外我还有几个问题问下
1.现在java-sdk有稳定的版本嘛
2.默认使用的是欧氏距离来计算相似度的，怎么使用余弦来计算呢？
3.有基于相似度的阈值来查询数据的api吗

**Reply by yhmo** [reply]
Java SDK 2.2.1是目前最新版本，应该没什么大的问题。
MetricType设置为"IP"就是余弦相似度，前提是所有的向量都要做归一化才能入库
目前没有相似度阈值查询的功能

**Reply by yangyangwen** [reply]
感谢感谢
但是我MetricType设置为"IP",也做了归一化(我理解的是取的值都是1~-1之间)，单但是算出来的score却很大
![image](https://user-images.githubusercontent.com/45092428/205277927-4e119b48-f44f-49e3-a48e-d81c04bf769f.png)

**Reply by yangyangwen** [reply]
另外有没有什么交流群呀，偶尔有问题需要咨询的呢

**Reply by yhmo** [reply]
score大于1，基本上都是L2计算的。要注意建索引和search时都要设置一样的metrictype，具体可以参考这个测试例子：https://github.com/milvus-io/milvus-sdk-java/blob/master/src/test/java/io/milvus/client/MilvusClientDockerTest.java

建索引时：
CreateIndexParam indexParam = CreateIndexParam.newBuilder().withMetricType(MetricType.IP)
搜索时：
SearchParam searchParam = SearchParam.newBuilder().withMetricType(MetricType.IP)

再加上每条向量都做归一化（每条向量的长度都为1），算出来的距离才是-1到1之间。

微信上有miluvs用户交流群，你先用微信搜索公众号"Zilliz"，然后关注这个公众号，你会看到里面很多技术文章，然后几乎每篇文章下面有个二维码，扫二维码并回复“技术交流”，机器人就会把你拉进群。注意里面不要发布和技术无关的广告，不然会被机器人踢的。

**Reply by yangyangwen** [reply]
好的，感谢
